pkg/controllers/management/k3supgrade/k3supgrade_handler.go

```diff
@@ -0,0 +1,174 @@
+		version := *cluster.Status.Version
+		cluster.Spec.K3sConfig = &v3.K3sConfig{
+			Version: version.String(),
+		}
+	}
+
+	if cluster.Spec.K3sConfig.Version == "" {
+		cluster.Spec.K3sConfig.Version = cluster.Status.Version.String()
+	}
+
+	isNewer, err := IsNewerVersion(cluster.Status.Version.GitVersion, cluster.Spec.K3sConfig.Version)
+	if err != nil {
+		return cluster, err
+	}
+
+	if !isNewer {
+		// cannot upgrade- version has not changed or proposed version is not newer than current version
+		return cluster, nil
+	}
+
+	// create or update k3supgradecontroller if necessary
+	if err = h.deployK3sUpgradeController(cluster.Name); err != nil {
+		return cluster, err
+	}
+
+	// deploy plans into downstream cluster
+	if err = h.deployPlans(cluster); err != nil {
+		return cluster, err
+	}
+
+	return cluster, nil
+}
+
+// deployK3sUpgradeController creates a rancher k3s upgrader controller if one does not exist.
+// Updates k3s upgrader controller if one exists and is not the newest available version.
+func (h *handler) deployK3sUpgradeController(clusterName string) error {
+	userCtx, err := h.manager.UserContext(clusterName)
+	if err != nil {
+		return err
+	}
+
+	projectLister := userCtx.Management.Management.Projects("").Controller().Lister()
+	systemProject, err := project.GetSystemProject(clusterName, projectLister)
+	if err != nil {
+		return err
+	}
+
+	templateID := k3sUpgraderCatalogName
+	template, err := h.templateLister.Get(namespace.GlobalNamespace, templateID)
+	if err != nil {
+		return err
+	}
+
+	latestTemplateVersion, err := utils.LatestAvailableTemplateVersion(template)
+	if err != nil {
+		return err
+	}
+
+	creator, err := h.systemAccountManager.GetSystemUser(clusterName)
+	if err != nil {
+		return err
+	}
+	systemProjectID := ref.Ref(systemProject)
+	_, systemProjectName := ref.Parse(systemProjectID)
+
+	nsClient := userCtx.Core.Namespaces("")
+	appProjectName, err := utils2.EnsureAppProjectName(nsClient, systemProjectName, clusterName, systemUpgradeNS, creator.Name)
+	if err != nil {
+		return err
+	}
+
+	appLister := userCtx.Management.Project.Apps("").Controller().Lister()
+	appClient := userCtx.Management.Project.Apps("")
+
+	latestVersionID := latestTemplateVersion.ExternalID
+
+	app, err := appLister.Get(systemProjectName, "rancher-k3s-upgrader")
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			return err
+		}
+
+		desiredApp := &v32.App{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "rancher-k3s-upgrader",
+				Namespace: systemProjectName,
+				Annotations: map[string]string{
+					"field.cattle.io/creatorId": creator.Name,
+				},
+			},
+			Spec: v32.AppSpec{
+				Description:     "Upgrade controller for k3s clusters",
+				ExternalID:      latestVersionID,
+				ProjectName:     appProjectName,
+				TargetNamespace: systemUpgradeNS,
+			},
+		}
+
+		// k3s upgrader doesn't exist yet, so it will need to be created
+		if _, err = appClient.Create(desiredApp); err != nil {
```

Comment:
Check the error isn't an alreadyExists error to save from races

Reply:
if it receives an already exists error then the handler should work on the next attempt